File: 228-summary-ranges/summary-ranges.py
# Optimal Code
class Solution:
    def summaryRanges(self, nums: list[int]) -> list[str]:
        # number of element in the array
        n = len(nums)
        # if the array is empty there are not ranges to return
        if n == 0:
            return []
        # stroe the final list of ranges
        result = []
        # start processing from the first elements
        i = 0
        # continue until every element is processed
        while i < n:
            # store the first number of the current ranges
            start = nums[i]
            # keep moving forward while the numbers are consecutive
            while (i + 1 < n and nums[i] + 1 == nums[i + 1]):
                i += 1
            # if the starting and ending numbers are diffrent create range such 0 -> 2
            if start != nums[i]:
                result.append(f"{start}->{nums[i]}")
            # otherwise the range contains only one numbers
            else:
                result.append(str(start))
            # move the next unprocessed numbers
            i += 1
        # return all summarized ranges
        return result

# Time Complexity : O(N)
# Space Complexity : O(N)

File: 228-summary-ranges/summary-ranges.cpp
// Optimal Code
class Solution {
public:
    vector<string> summaryRanges(vector<int>& nums) {
        // number of elements in the array
        int n = nums.size();
        // if the array is empty, there are no ranges to return
        if (n == 0) {
            return {};
        }
        // store the final list of ranges
        vector<string> result;
        // start processing from the first element
        int i = 0;
        // continue until every element is processed
        while (i < n) {
            // store the first number of the current range
            int start = nums[i];
            // keep moving forward while the numbers are consecutive
            while (i + 1 < n && nums[i] + 1 == nums[i + 1]) {
                i++;
            }
            // if the starting and ending numbers are different create a range such as "0->2"
            if (start != nums[i]) {
                result.push_back(
                    to_string(start) + "->" + to_string(nums[i])
                );
            }
            // otherwise, the range contains only one number
            else {
                result.push_back(to_string(start));
            }
            // move to the next unprocessed number
            i++;
        }
        // return all summarized ranges
        return result;
    }
};

// Time Complexity : O(N)
// Space Complexity : O(N)